File: day1/main.ts
import fs from "node:fs";
function start() {
	const input = fs.readFileSync("input.txt", { encoding: "ascii" });
	const lines = input.split(/\r?\n/);
	lines.pop();
	const arr1: number[] = [];
	const arr2: number[] = [];
	lines.forEach((line) => {
		const nums = line.split(/\s+/);
		arr1.push(parseInt(nums[0]));
		arr2.push(parseInt(nums[1]));
	});
	console.log(answer1(arr1, arr2));
	console.log(answer2(arr1, arr2));
}
function answer1(l1: number[], l2: number[]): number {
	l1.sort();
	l2.sort();
	let sum = 0;
	for (const idx in l1) {
		sum += Math.abs(l1[idx] - l2[idx]);
	}
	return sum;
}
function answer2(l1: number[], l2: number[]): number {
	const set = new Set(l1.sort());
	return set.values().reduce(
		(prev, current) =>
			prev + current * l2.filter((num) => num === current).length,
		0,
	);
}
start();

File: day1/main.rb
# frozen_string_literal:true

def answer1(arr1, arr2)
  arr1.sort!
  arr2.sort!

  sum = 0
  arr1.size.times do |idx|
    sum += (arr1[idx] - arr2[idx]).abs
  end

  sum
end

def answer2(arr1, arr2)
  arr1.to_set.reduce(0) { |acc, x| acc + x * arr2.count(x) }
end

def start
  arr1 = []
  arr2 = []
  File.open('input.txt').readlines.each do |line|
    nums = line.split('   ')
    arr1.push(nums[0].to_i)
    arr2.push(nums[1].to_i)
  end

  puts answer1 arr1, arr2
  puts answer2 arr1, arr2
end

start
